File: ddm_checks/compare_tapas_wiener_RWiener.R
# script to be called from MATLAB; wraps around dwiener
args <- commandArgs(trailingOnly=TRUE)
alpha <- if (length(args) >= 1) as.numeric(args[1]) else 1
tau   <- if (length(args) >= 2) as.numeric(args[2]) else 0.1
beta  <- if (length(args) >= 3) as.numeric(args[3]) else 0.5
delta <- if (length(args) >= 4) as.numeric(args[4]) else 1
output_file <- if (length(args) >= 5) args[5] else 'check_RWiener_log_p.csv'

library(RWiener)

df <- data.frame(rt = seq(from = (tau + 1e-6), to = 10, length.out = 1000))
df$p_upper <- NA
df$p_lower <- NA
for (i in seq_len(nrow(df))) {
  rt_i <- df$rt[i]
  df$p_upper[i] <- dwiener(rt_i, alpha=alpha, tau=tau, beta=beta, delta=delta, resp="upper")
  df$p_lower[i] <- dwiener(rt_i, alpha=alpha, tau=tau, beta=beta, delta=delta, resp="lower")
}
write.csv(df, output_file, row.names = FALSE)

File: ddm_checks/compare_tapas_wiener_RWiener.m
function [temp_file_dir, df] = compare_tapas_wiener_RWiener(alpha, tau, beta, delta, do_vis)

% Bowen Xiao 20250606; updated for tapas_wiener in 20251202
% calls an Rscript to compute probability density of a range of RTs
% at a given combination of 4par DDM params
% then, compare the output of RWiener::dwiener to tapas_wfpt

if nargin < 5, do_vis = true; end

%% run RWiener::dwiener
% Create unique filename to avoid race conditions
temp_file_dir = sprintf('check_RWiener_log_p_%d.csv', randi(999999));
% Use %.17g or %.17f to ensure full double precision is passed to R
cmd = sprintf('/usr/local/bin/Rscript compare_tapas_wiener_RWiener.R %.17f %.17f %.17f %.17f %s', ...
              alpha, tau, beta, delta, temp_file_dir);
system(cmd);

%% load R results and find the same lpdf in MATLAB
df = readtable(temp_file_dir);

% handling of small prob - sync with wiener_lpdf!!!
df.log_p_upper = log(max(df.p_upper, eps));
df.log_p_lower = log(max(df.p_lower, eps));

my_p_upper = nan(size(df.rt));
my_p_lower = nan(size(df.rt));

for i = 1:height(df)
    % note that there is an upper-lower switch...
    my_p_upper(i) = tapas_wiener_pdrts(df.rt(i), alpha, tau, 1-beta, -delta);
    my_p_lower(i) = tapas_wiener_pdrts(df.rt(i), alpha, tau, beta, delta);
end

% handling of small prob
df.tapas_log_p_upper = log(max(my_p_upper, eps));
df.tapas_log_p_lower = log(max(my_p_lower, eps));

% compute absolute difference
upper_err = max(abs(df.log_p_upper - df.tapas_log_p_upper));
lower_err = max(abs(df.log_p_lower - df.tapas_log_p_lower));

writetable(df,temp_file_dir);

%% visualise
if do_vis
    subplot(2,2,1)
    pal_scat_ref_corr(df.log_p_upper,my_p_upper)
    xlabel('RWiener pdf Upper')

    subplot(2,2,2)
    pal_scat_ref_corr(df.log_p_lower,my_p_lower)
    xlabel('RWiener pdf Lower')

    subplot(2,2,3)
    plot(df.rt, exp(my_p_upper),'-')
    hold on
    plot(df.rt, exp(df.log_p_upper),'--')
    hold off
    xlabel('Reaction time')
    ylabel('Probability density')
    title(sprintf('\\bfα=%.2f, τ=%.2f, β=%.2f, δ=%.2f', alpha, tau, beta, delta))

    subplot(2,2,4)
    plot(df.rt, exp(my_p_lower),'-')
    hold on
    plot(df.rt, exp(df.log_p_lower),'--')
    hold off
    xlabel('Reaction time')
    ylabel('Probability density')
    title(sprintf('\\bfα=%.2f, τ=%.2f, β=%.2f, δ=%.2f', alpha, tau, beta, delta))
    
    sgtitle(sprintf('Max error for upper: %.0e\nMax error for lower: %.0e', upper_err, lower_err))
end

end